### DISCUSSION TITLE
Live Objects Enhancements

### DISCUSSION THREAD

**Discussion by indygreg** [async-profiler/async-profiler#1012] [Ideas]
I think the utility of `--live` could be improved to make it easier to find leaks.

From my reading of the source code the way `--live` mode works today is as follows:

* Profiler maintains a ring buffer of 1024 weak references to sampled allocated objects.
* When the ring is full, new adds are paused.
* When a GC occurs, new adds are allowed and each slot is examined to see if the ref is still valid. If not, the entry is discarded and the slot reused. This occurs until the ring is full again or another GC occurs.
* On JFR dump, the ref is checked and non-live objects are filtered out.

The objects in `LiveObject` events are only for live objects. This results in an accurate picture of, well, live objects.

But the advertised objects consist of "old" and "young" objects, with varying degrees of tenuring. (It isn't clear to me what `NewLocalRef()` - the function used for filtering if we emit a `LiveObject` JFR event - returns for objects that would be collected by the next GC. I think it must return non-null if a GC hasn't ran yet, right? If so, then doesn't `LiveObject` contain references to _zombie_ objects that would be collected imminently?)

The purported purpose of `--live` mode is to facilitate leak analysis. I think it would be much more effective at that role if it grew a few features. Here are some of my ideas.

## GC Count Threshold

We allow specifying how many GC operations an object must survive for it to be emitted.

We maintain a circular buffer of ticks of N most recent GCs. We filter out objects newer than our GC threshold.

This allows us to aggressively filter out _young_ objects that aren't _sufficiently tenured_. A user could set/increase this to find objects that are tenuring. IMO this is a much better filter for finding leaking objects.

## Flushing / Stopping When Ring is Full

For JVMs that do have a memory leak and no amount of GCs will free objects, eventually the weakref ring will become full and GCs won't free any capacity.

I think it would be cool if there were a mode - say `--live-stop-at-full` - where the profiler detects when GCs fail to collect any of its tracked live references and this triggers the profiler to stop and dump live objects.

A problem today is that when you are running in live objects mode you don't know how long to let it run. Too short and you don't find the leak. Too long and you are wasting time. Having the profiler self-terminate when leaks have been detected could be far more usable for end-users.

What do you think?

**Comment by apangin** [comment]
Thank you for your suggestions. We do plan to improve live object profiling, especially in relation to continuous profiling.
Do you consider contributing pull request(s) for the proposed features?

**Comment by indygreg** [comment]
I could potentially contribute improvements. But I'm a bit pressed for time and don't have guarantees when that might be. I might be able to cobble together an initial implementation this weekend or next.

It would help if I could have some guidance on which technical solutions / features are desirable.

I was looking at the JVMTI docs and noticed that `GarbageCollectionStart` and `GarbageCollectionFinish` only fire during pauses and there's a caveat: `This means that some collectors will never generate these events.` That's a pretty big caveat! Although practically speaking I'm unsure if it is a problem. `SvcGCMarker` is the symbol/function you want to search for which triggers the JVMTI GC event. And I see G1, Parallel, Shenandoah, and Z all calling it in the OpenJDK 21 code. That means Epsilon and Serial potentially leave it out. Since the JVMTI GC events may not be reliable, I suppose that means we need to make any GC-based filtering of live objects a conditional feature since it may not _just work_. As a user convenience we could sniff the in-use GC and opt in if a `GarbageCollectionStart` compatible GC is in use - if we wanted the extra code complexity.
